There are rumors of a bug with SuccessOnSelected

In the rumor mill there are whisperings of special characters not being supported in children names by `SuccessOnSelected`.

```
<py_trees.composites.Parallel name="Move and Scan"
  policy="$(py_trees.common.ParallelPolicy.SuccessOnSelected([camera_val], synchronise=False))">
<py_trees.behaviours.SetBlackboardVariable
  name="Write camera type"
  variable_name="/perception/camera"
  variable_value="rgb"
  overwrite="true"
/>
<py_trees.behaviours.CheckBlackboardVariableValue
  name="camera_val"
  check="$(py_trees.common.ComparisonExpression('/perception/camera', value='rgb', operator=operator.is_))"/>
</py_trees.composites.Parallel>
```

this one works fine, but if you change it to having space between the words in the behavior name it goes like this

----

```
<py_trees.composites.Parallel name="Move and Scan"
  policy="$(py_trees.common.ParallelPolicy.SuccessOnSelected([camera val], synchronise=False))">
<py_trees.behaviours.SetBlackboardVariable
  name="Write camera type"
  variable_name="/perception/camera"
  variable_value="rgb"
  overwrite="true"
/>
<py_trees.behaviours.CheckBlackboardVariableValue
  name="camera val"
  check="$(py_trees.common.ComparisonExpression('/perception/camera', value='rgb', operator=operator.is_))"/>
</py_trees.composites.Parallel>
```

> [INFO] [launch]: All log files can be found below /home/ros/.ros/log/2025-08-27-15-01-11-114445-31T32Z2-2982
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [behavior_tree-1]: process started with pid [2983]
[behavior_tree-1] [INFO] [1756306871.877762867] [irb1200_tree]: Setting log level to info
[behavior_tree-1] [WARN] [1756306871.886306470] [irb1200_tree]: No data in trigger config, no triggers added to robot
[behavior_tree-1] [ERROR] [1756306871.891464771] [BTParser]: Couldn't parse tree: Couldn't read function parameters: py_trees.common.ParallelPolicy.SuccessOnSelected([camera val], synchronise=False)
[INFO] [behavior_tree-1]: process has finished cleanly [pid 2983]

----

and finally with double quotes like you said it does work fine.

```
<py_trees.composites.Parallel name="Move and Scan"
  policy="$(py_trees.common.ParallelPolicy.SuccessOnSelected(['camera val'], synchronise=False))">
<py_trees.behaviours.SetBlackboardVariable
  name="Write camera type"
  variable_name="/perception/camera"
  variable_value="True"
  overwrite="true"
/>
<py_trees.behaviours.CheckBlackboardVariableValue
  name="'camera val'"
  check="$(py_trees.common.ComparisonExpression('/perception/camera', value=True, operator=operator.is_))"/>
</py_trees.composites.Parallel>
```

```
********************************************************************************
*                                     Tick 51                                     
********************************************************************************
/_/ Move and Scan [✓]
    --> Write camera type [✓]
    --> 'camera val' [✓] -- '/perception/camera' comparison succeeded [v: True][e: True]
--------------------------------------------------------------------------------
Blackboard Data
    /perception/camera : True
Blackboard Activity Stream
    /perception/camera : WRITE         | Write camera type | → True
    /perception/camera : READ          | 'camera val'      | ← True
```

----

on a side note: I changed for this last example above the variable to be a `bool` because the comparison failed with strings and I wasn't sure if it's another bug or me not correctly writing one of the params. with this example below having the original `str` value written in blackboard

```
<py_trees.composites.Parallel name="Move and Scan"
  policy="$(py_trees.common.ParallelPolicy.SuccessOnSelected(['camera val'], synchronise=False))">
<py_trees.behaviours.SetBlackboardVariable
  name="Write camera type"
  variable_name="/perception/camera"
  variable_value="rgb"
  overwrite="true"
/>
<py_trees.behaviours.CheckBlackboardVariableValue
  name="'camera val'"
  check="$(py_trees.common.ComparisonExpression('/perception/camera', value='rgb', operator=operator.is_))"/>
</py_trees.composites.Parallel>
```

got comparison failed

```
********************************************************************************
*                                     Tick 543                                    
********************************************************************************

/_/ Move and Scan [✕]
    --> Write camera type [✓]
    --> 'camera val' [✕] -- '/perception/camera' comparison failed [v: rgb][e: rgb]

--------------------------------------------------------------------------------

Blackboard Data
    /perception/camera : rgb

Blackboard Activity Stream
    /perception/camera : WRITE         | Write camera type | → rgb
    /perception/camera : READ          | 'camera val'      | ← rgb
```

and if I lose the quotes arounf rgb in `CheckBlackboardVariableValue` like this

```
<py_trees.composites.Parallel name="Move and Scan"
  policy="$(py_trees.common.ParallelPolicy.SuccessOnSelected(['camera val'], synchronise=False))">
<py_trees.behaviours.SetBlackboardVariable
  name="Write camera type"
  variable_name="/perception/camera"
  variable_value="rgb"
  overwrite="true"
/>
<py_trees.behaviours.CheckBlackboardVariableValue
  name="'camera val'"
  check="$(py_trees.common.ComparisonExpression('/perception/camera', value=rgb, operator=operator.is_))"/>
</py_trees.composites.Parallel>
```

then I get

>[INFO] [behavior_tree-1]: process started with pid [7156]
[behavior_tree-1] [INFO] [1756307873.031845676] [irb1200_tree]: Setting log level to info
[behavior_tree-1] [WARN] [1756307873.074462885] [irb1200_tree]: No data in trigger config, no triggers added to robot
[behavior_tree-1] [ERROR] [1756307873.081305302] [BTParser]: Couldn't parse tree: name 'rgb' is not defined

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -1,6 +1,10 @@
 Changelog for package py_trees_parser
 
 .. This is only a rough description of the main changes of the repository
+0.7.1 (2025-08-27)
+------------------
+* SuccessOnSlected handling of special characters
+
 0.7.0 (2025-03-28)
 ------------------
 * Add support for SuccessOnSelected parallel policy

--- a/README.md
+++ b/README.md
@@ -113,6 +113,9 @@ parser = BTParser(xml_file)
 behavior_tree = parser.parse()
 ```
 
+note: The names given to children above would use an underscore, i.e. `_` for
+      any special characters, e.g. " ", "$", "@", "!", etc.
+
 ### Using Your Own Behaviors
 
 The xml parser can use any behavior, whether it is part of `py_trees`, `py_trees_ros`,
@@ -147,7 +150,6 @@ then you would include the behaviors in `my_behavior.py` in the following way:
 
 The path to this can be shortened by including the class in `__init__.py`.
 
-
 ### Sub-Trees
 
 It is possible to include sub-trees in the xml file containing a behavior tree.
@@ -171,6 +173,7 @@ is possible to use python to determine the path like so:
       include="$(os.path.join(ament_index_python.packages.get_package_share_directory('my_package'), 'tree', 'subtree.xml'))" />
 </py_trees.composites.Parallel>
 ```
+
 #### Arguments
 
 It is also possible to use arguments for subtrees. The syntax of which looks like

--- a/package.xml
+++ b/package.xml
@@ -13,6 +13,7 @@
   <exec_depend>pynput-pip</exec_depend>
   <exec_depend>python3-setuptools</exec_depend>
 
+  <test_depend>action_msgs</test_depend>
   <test_depend>ament_copyright</test_depend>
   <test_depend>python3-pytest</test_depend>
 

--- a/package.xml
+++ b/package.xml
@@ -2,7 +2,7 @@
 <?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
 <package format="3">
   <name>py_trees_parser</name>
-  <version>0.7.0</version>
+  <version>0.7.1</version>
   <description>A py_trees xml parser</description>
   <maintainer email="e.l.f.foster@tudelft.nl">Erich L Foster</maintainer>
   <license>Apache-2.0</license>

--- a/py_trees_parser/parser.py
+++ b/py_trees_parser/parser.py
@@ -502,7 +502,12 @@ class BTParser:
         elif on_selected is not None:
             self.logger.debug("Found SuccessOnSelected in parameters")
 
-            selection = [child for child in children if child.name in on_selected]
+            selection = [
+                child
+                for child in children
+                if "".join(map(lambda char: char if char.isalnum() else "_", child.name))
+                in on_selected
+            ]
             if len(selection) == 0:
                 raise ValueError("List of children for SuccessOnSelected is empty")
 

--- a/py_trees_parser/parser.py
+++ b/py_trees_parser/parser.py
@@ -502,11 +502,17 @@ class BTParser:
         elif on_selected is not None:
             self.logger.debug("Found SuccessOnSelected in parameters")
 
+            def replace_special(char: str):
+                if char.isalnum():
+                    return char
+
+                self.logger.warning(f"Found special {char} in behavior name, replacing with '_'")
+                return "_"
+
             selection = [
                 child
                 for child in children
-                if "".join(map(lambda char: char if char.isalnum() else "_", child.name))
-                in on_selected
+                if "".join(map(replace_special, child.name)) in on_selected
             ]
             if len(selection) == 0:
                 raise ValueError("List of children for SuccessOnSelected is empty")

--- a/setup.py
+++ b/setup.py
@@ -20,7 +20,7 @@ package_name = "py_trees_parser"
 
 setup(
     name=package_name,
-    version="0.7.0",
+    version="0.7.1",
     packages=find_packages(exclude=["test"]),
     data_files=[
         ("share/ament_index/resource_index/packages", [os.path.join("resource", package_name)]),

--- a/test/data/test_on_selected.xml
+++ b/test/data/test_on_selected.xml
@@ -1,0 +1,13 @@
+<py_trees.composites.Parallel name="Move and Scan"
+  policy="$(py_trees.common.ParallelPolicy.SuccessOnSelected([__mera_va_ue__check_], synchronise=False))">
+  <py_trees.behaviours.SetBlackboardVariable
+    name="Write camera type"
+    variable_name="/perception/camera"
+    variable_value="rgb"
+    overwrite="true"
+  />
+  <py_trees.behaviours.CheckBlackboardVariableValue
+    name="¢@mera va!ue,_check?"
+    check="$(py_trees.common.ComparisonExpression('/perception/camera', value='rgb', operator=operator.is_))"
+  />
+</py_trees.composites.Parallel>

--- a/test/test_parser.py
+++ b/test/test_parser.py
@@ -224,7 +224,8 @@ def test_SuccessOnSelected_parsing(child_key, synch_key, children, synch):
     assert "policy" not in node.attrib
 
 
-def test_SuccessOnSelected_speacial_chars_in_name(setup_parser):
+def test_SuccessOnSelected_special_chars_in_name(setup_parser):
+    """Test SuccessOnSelected can handle special characters correctly."""
     tree_file = "test_on_selected.xml"
     root = setup_parser(tree_file)
     assert isinstance(root.policy, py_trees.common.ParallelPolicy.SuccessOnSelected)

--- a/test/test_parser.py
+++ b/test/test_parser.py
@@ -224,6 +224,14 @@ def test_SuccessOnSelected_parsing(child_key, synch_key, children, synch):
     assert "policy" not in node.attrib
 
 
+def test_SuccessOnSelected_speacial_chars_in_name(setup_parser):
+    tree_file = "test_on_selected.xml"
+    root = setup_parser(tree_file)
+    assert isinstance(root.policy, py_trees.common.ParallelPolicy.SuccessOnSelected)
+    assert len(root.policy.children) != 0
+    assert root.policy.children[0].name == "¢@mera va!ue,_check?"
+
+
 def test_ParallelPolicy(setup_parser):
     """Test parallel policies are handled correctly."""
     tree_file = "test_parallel_policy.xml"

